[livestreamfails] Add new extractor

nomevi opened the pull request:
<!--
# Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information
Add an extractor for [Livestreamfails](https://livestreamfails.com/), a Twitch clip mirror site for [r/livestreamfail](https://reddit.com/r/LivestreamFail/).

Ashish0804 commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,50 @@
+            'display_id': 'ConcernedLitigiousSalmonPeteZaroll-O8yo9W2L8OZEKhV2',
+            'title': 'Streamer jumps off a trampoline at full speed',
+            'creator': 'paradeev1ch',
+            'thumbnail': 'https://livestreamfails-image-prod.b-cdn.net/image/3877b1d38db083fa25c82685bbaf645637e575ea.png',
```
better to use regex here

nomevi replied:
To what extent? The [example](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-support-for-a-new-site) has regex for the thumbnail but not the title.

Lesmiscore replied:
this review points to `thumbnail` key

nomevi replied:
My guess is that this is because the image should be able to change URL as long as it is still results in a 200. Is that correct?

pukkandan replied:
Yes. If you know the URL doesn't change over time, you can keep the current test. Otherwise, use a regex to future-proof

Ashish0804 commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,50 @@
+        'info_dict': {
+            'id': '139200',
+            'ext': 'mp4',
+            'display_id': 'ConcernedLitigiousSalmonPeteZaroll-O8yo9W2L8OZEKhV2',
```
don't see how this is useful

nomevi replied:
`id` and `ext` are from the [example](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-support-for-a-new-site). Should I remove them?

And as for the `display_id`: Twitch has its own IDs for clips (used like `https://clips.twitch.tv/ConcernedLitigiousSalmonPeteZaroll-O8yo9W2L8OZEKhV2`). I figured I should include that since the regular ID is only valid for livestreamfails.com. Should I not include the Twitch ID?

Ashish0804 commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,50 @@
+    }]
+
+    def _real_extract(self, url):
+        result = {}
+        result['id'] = self._match_id(url)
+
+        # https://livestreamfails.com/clip/id uses https://api.livestreamfails.com/clip/ to fetch the video metadata
+        # Use the same endpoint here to avoid loading and parsing the provided page (which requires JS)
+        apiResponse = json.loads(self._download_webpage('https://api.livestreamfails.com/clip/' + result['id'], result['id']))
+
+        # Twitch ID of clip
+        result['display_id'] = apiResponse.get('sourceId')
+
+        # Get the input timestamp (test case gives 2022-06-26T19:29:45.515Z)
+        result['timestamp'] = apiResponse.get('createdAt')
+        if(result.get('timestamp')):
+            # Parse it into a struct_time
+            result['timestamp'] = time.strptime(result['timestamp'], '%Y-%m-%dT%H:%M:%S.%fZ')
+            # Convert the struct_time to a UNIX timestamp while ignoring the local timezone attached by time.strptime()
+            result['timestamp'] = calendar.timegm(result['timestamp'])
+
+        # Other fields
+        result['url'] = 'https://livestreamfails-video-prod.b-cdn.net/video/' + apiResponse.get('videoId')
+        result['title'] = apiResponse.get('label')
+        result['creator'] = traverse_obj(apiResponse, ('streamer', 'label'))
+        result['thumbnail'] = 'https://livestreamfails-image-prod.b-cdn.net/image/' + apiResponse.get('imageId')
+
+        return result
```
bruh

pukkandan reviewed (changes requested):
1. Follow basic Python coding conventions. Eg: snake_case for variables
1. All the comments are useless
1. https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#use-convenience-conversion-and-parsing-functions
5. Don't create empty dict and update. You can directly return:
```
return {
    'display_id': apiResponse.get('sourceId'),
    ...
}
```

nomevi commented:
> 1. Follow basic Python coding style. Eg: snake_case for variables

My bad, I'm not used to Python. Fixed.

> 2. All the comments are useless

Removed.

> 3. https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#use-convenience-conversion-and-parsing-functions

Using `unified_timestamp` now.

> 4. Don't create empty dict and update.

I did that to fit all the comments. Fixed now.

pukkandan commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,37 @@
+            'id': video_id,
+            'display_id': api_response.get('sourceId'),
+            'timestamp': unified_timestamp(api_response.get('createdAt')),
+            'url': 'https://livestreamfails-video-prod.b-cdn.net/video/' + api_response.get('videoId'),
```
```suggestion
            'url': f'https://livestreamfails-video-prod.b-cdn.net/video/{api_response["videoId"]}',
```

pukkandan commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,37 @@
+            'url': 'https://livestreamfails-video-prod.b-cdn.net/video/' + api_response.get('videoId'),
+            'title': api_response.get('label'),
+            'creator': traverse_obj(api_response, ('streamer', 'label')),
+            'thumbnail': 'https://livestreamfails-image-prod.b-cdn.net/image/' + api_response.get('imageId')
```
Use `format_field` so that `None` is handled correctly

pukkandan replied:
this is unresolved

pukkandan commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,37 @@
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        api_response = json.loads(self._download_webpage('https://api.livestreamfails.com/clip/' + video_id, video_id))
```
_download_json
```suggestion
        api_response = self._download_json(f'https://api.livestreamfails.com/clip/{video_id}', video_id)
```

pukkandan reviewed (changes requested):
u sure the download test still passes?

nomevi commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,37 @@
+            'display_id': 'ConcernedLitigiousSalmonPeteZaroll-O8yo9W2L8OZEKhV2',
+            'title': 'Streamer jumps off a trampoline at full speed',
+            'creator': 'paradeev1ch',
+            'thumbnail': r're:^https?://.*\..*$',
```
I shouldn't have removed `upload_date`. I removed the suffix from the thumbnail URL to account for the developer switching formats.

nomevi commented:
Both the test and linting are good now.

pukkandan commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,37 @@
+from .common import InfoExtractor
+from ..utils import (
+    traverse_obj,
+    unified_timestamp,
+)
```
```suggestion
from ..utils import traverse_obj, unified_timestamp
```

pukkandan commented on yt_dlp/extractor/livestreamfails.py:
```diff
@@ -0,0 +1,34 @@
+            'url': f'https://livestreamfails-video-prod.b-cdn.net/video/{api_response["videoId"]}',
+            'title': api_response.get('label'),
+            'creator': traverse_obj(api_response, ('streamer', 'label')),
+            'thumbnail': format_field(api_response.get('imageId'), None, "https://livestreamfails-image-prod.b-cdn.net/image/%s")
```
```suggestion
            'thumbnail': format_field(api_response, 'imageId', 'https://livestreamfails-image-prod.b-cdn.net/image/%s')
```
